modules/renter/hostdb/hosttree/hosttree.go

```diff
@@ -0,0 +1,258 @@
+		n.entry = entry
+		n.taken = true
+		n.weight = entry.Weight
+		nodesAdded = 0
+		newnode = n
+		return
+	}
+
+	n.weight = n.weight.Add(entry.Weight)
+
+	// If the current node is empty, add the entry but don't increase the
+	// count.
+	if !n.taken {
+		n.taken = true
+		n.entry = entry
+		newnode = n
+		return
+	}
+
+	// Insert the element into the lest populated side.
+	if n.left == nil {
+		n.left = createNode(n, entry)
+		nodesAdded = 1
+		newnode = n.left
+	} else if n.right == nil {
+		n.right = createNode(n, entry)
+		nodesAdded = 1
+		newnode = n.right
+	} else if n.left.count <= n.right.count {
+		nodesAdded, newnode = n.left.recursiveInsert(entry)
+	} else {
+		nodesAdded, newnode = n.right.recursiveInsert(entry)
+	}
+
+	n.count += nodesAdded
+	return
+}
+
+// nodeAtWeight grabs an element in the tree that appears at the given weight.
+// Though the tree has an arbitrary sorting, a sufficiently random weight will
+// pull a random element. The tree is searched through in a post-ordered way.
+func (n *node) nodeAtWeight(weight types.Currency) (*node, error) {
+	// Sanity check - weight must be less than the total weight of the tree.
+	if weight.Cmp(n.weight) > 0 {
+		return nil, ErrWeightTooHeavy
+	}
+
+	// Check if the left or right child should be returned.
+	if n.left != nil {
+		if weight.Cmp(n.left.weight) < 0 {
+			return n.left.nodeAtWeight(weight)
+		}
+		weight = weight.Sub(n.left.weight) // Search from the 0th index of the right side.
+	}
+	if n.right != nil && weight.Cmp(n.right.weight) < 0 {
+		return n.right.nodeAtWeight(weight)
+	}
+
+	// Should we panic here instead?
+	if !n.taken {
+		return nil, ErrNilEntry
+	}
+
+	// Return the root entry.
+	return n, nil
+}
+
+// remove takes a node and removes it from the tree by climbing through the
+// list of parents. remove does not delete nodes.
+func (n *node) remove() {
+	n.weight = n.weight.Sub(n.entry.Weight)
+	n.taken = false
+	current := n.parent
+	for current != nil {
+		current.weight = current.weight.Sub(n.entry.Weight)
+		current = current.parent
+	}
+}
+
+// Insert inserts the entry provided to `entry` into the host tree. Insert will
+// return an error if the input host already exists.
+func (ht *HostTree) Insert(entry *HostEntry) error {
+	if _, exists := ht.hosts[entry.PublicKey.String()]; exists {
+		return ErrHostExists
+	}
+
+	_, node := ht.recursiveInsert(entry)
+
+	ht.hosts[entry.PublicKey.String()] = node
+	return nil
+}
+
+// Remove removes the host with the public key provided by `pk`.
+func (ht *HostTree) Remove(pk types.SiaPublicKey) error {
+	node, exists := ht.hosts[pk.String()]
+	if !exists {
+		return ErrNoSuchHost
+	}
+	node.remove()
+	delete(ht.hosts, pk.String())
+
+	return nil
+}
+
+// Modify updates a host entry at the given public key, replacing the old entry
+// with the entry provided by `newEntry`.
+func (ht *HostTree) Modify(pk types.SiaPublicKey, newEntry *HostEntry) error {
```

Comment:
why does this function need to take a public key, if the public key is also in `newEntry`?